sklearn/semi_supervised/_self_training.py

```diff
@@ -0,0 +1,318 @@
+import numpy as np
+
+from ..base import MetaEstimatorMixin, clone
+from ..utils.validation import check_X_y, check_array, check_is_fitted
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils import safe_mask
+from ..exceptions import ConvergenceWarning
+from ..utils.metaestimators import _BaseComposition
+import warnings
+
+__all__ = ["SelfTrainingClassifier"]
+
+# Authors: Oliver Rausch   <rauscho@ethz.ch>
+#          Patrice Becker  <beckerp@ethz.ch>
+# License: BSD 3 clause
+
+
+def _validate_estimator(estimator):
+    """Make sure that an estimator implements the necessary methods."""
+    if not hasattr(estimator, "predict_proba"):
+        name = type(estimator).__name__
+        msg = "base_classifier ({}) should implement predict_proba!"
+        msg = msg.format(name)
+        raise ValueError(msg)
+
+
+class SelfTrainingClassifier(MetaEstimatorMixin, _BaseComposition):
+    """Self-training classifier
+
+    This class allows a given supervised classifier to function as a
+    semi-supervised classifier, allowing it to learn from unlabeled data. It
+    does this by iteratively predicting labels for the unlabeled data.
+
+    Read more in the :ref:`User Guide <self_training>`.
+
+    Parameters
+    ----------
+    base_classifier : estimator object
+        An estimator object implementing ``fit`` and ``predict_proba``.
+        Invoking the ``fit`` method will fit a clone of the passed estimator,
+        which will be stored in the ``base_classifier_`` attribute.
+
+    threshold : float, optional (default=0.75)
+        The decision threshold. If the ``base_classifier`` makes a prediction
+        with a ``predict_proba`` above this threshold, it will be added to the
+        labeled dataset.
+        Should be in [0, 1).
+
+    max_iter : int or ``None``, optional (default=10)
+        Maximum number of iterations allowed. Should be greater than or equal
+        to 0. If it is ``None``, the classifier will continue to predict labels
+        until all unlabeled samples have been labeled. In this case, be aware
+        that the fit may never terminate.
+
+    n_iter_no_change: int, optional (default=3)
+        If not `None`, early stopping will be applied. In this case, the
+        classifier will count the amount of new labels in each iteration. If
+        no new labels are added to the training set for `n_iter_no_change`
+        iterations, the classifier will stop fitting early.
```

Comment:
In literature, if early stopping is used, it is usually with only one iteration. I added a `n_iter_no_change` parameter because it seemed like a reasonable generalization that might be useful on nondeterministic classifiers. I don't have enough experience with this usecase to say when it is effective to raise the parameter above 1 in practice.

Reply:
It would be great to give it a try (in a temporary gist), e.g. with `SGDClassifier` for instance.

Reply:
Yes, I think always stopping when there's no change makes sense. We don't need a parameter to control it. Nor do we need to worry about max_iter=None never terminating.

Reply:
Early stopping is now always applied.